telluric/georaster.py

```diff
@@ -101,21 +103,97 @@ def merge_all(rasters, roi=None, dest_resolution=None, merge_strategy=MergeStrat
        When roi is provided the ul_corner, shape and crs are ignored
     """
     if dest_resolution is None:
-        dest_resolution = rasters[0].resolution()
+        dest_resolution = rasters[0].res_xy()
 
     # Create empty raster
     empty = GeoRaster2.empty_from_roi(
         roi, resolution=dest_resolution, band_names=rasters[0].band_names,
         dtype=rasters[0].dtype, shape=shape, ul_corner=ul_corner, crs=crs)
 
-    projected_rasters = [_prepare_other_raster(empty, raster) for raster in rasters]
-    projected_rasters = [raster for raster in projected_rasters if raster is not None]
-    func = partial(_merge, merge_strategy=merge_strategy)  # type: Callable[[_Raster, _Raster], _Raster]
-    raster = reduce(func, projected_rasters, empty)
-    return empty.copy_with(image=raster.image, band_names=raster.band_names)
+    # Create a list of single band rasters
+    all_band_names, projected_rasters = _prepare_rasters(rasters, merge_strategy, empty)
+
+    if all_band_names:
+        # Merge common bands
+        projected_rasters = _merge_common_bands(projected_rasters)
+
+        # Merge all bands
+        merge_union = partial(_merge, merge_strategy=MergeStrategy.UNION)  # type: Callable[[_Raster, _Raster], _Raster]
+        raster = reduce(merge_union, projected_rasters)
+
+        return empty.copy_with(image=raster.image, band_names=raster.band_names)
+
+    else:
+        raise ValueError("result contains no bands, use another merge strategy")
+
+
+def _merge_common_bands(rasters):
+    # type: (List[_Raster]) -> List[_Raster]
+    """Combine the common bands.
+
+    """
+    merge_intersect = partial(_merge,
+                              merge_strategy=MergeStrategy.INTERSECTION)  # type: Callable[[_Raster, _Raster], _Raster]
+
+    # Compute band order
+    all_bands = IndexedSet([rs.band_names[0] for rs in rasters])
+
+    def key(rs):
+        return all_bands.index(rs.band_names[0])
+
+    rasters_final = []  # type: List[_Raster]
+    for band_name, rasters_group in groupby(sorted(rasters, key=key), key=key):
+        rasters_final.append(reduce(merge_intersect, rasters_group))
+
+    return rasters_final
+
+
+def _prepare_rasters(rasters, merge_strategy, first):
+    # type: (List[GeoRaster2], MergeStrategy, GeoRaster2) -> Tuple[IndexedSet[str], List[_Raster]]
+    """Prepares the rasters according to the requested ROI and merge strategy.
+
+    """
+    # Create list of prepared rasters
+    all_band_names = IndexedSet(first.band_names)
+    projected_rasters = []  # type: List[GeoRaster2]
+    for raster in rasters:
+        projected_raster = _prepare_other_raster(first, raster)
+
+        # Modify the bands only if an intersecting raster was returned
+        if projected_raster:
+            if merge_strategy is MergeStrategy.INTERSECTION:
+                all_band_names.intersection_update(projected_raster.band_names)
```

Comment:
in merge strategy intersection, if I read the code correctly, there is a case you decide not to add the raster bands because there is not intersection with the bands, but the raster will be added to the projected_rasters in line 169, is that okay?

Reply:
If `all_band_names` becomes the empty set because of the last intersection, you mean?